src/BotPlutusInterface/Balance.hs

```diff
@@ -148,7 +151,12 @@ balanceTxIO' balanceCfg pabConf ownPkh unbalancedTx =
           else hoistEither $ addSignatories ownPkh privKeys requiredSigs tx
 
       -- Balance the tx
-      (balancedTx, minUtxos) <- balanceTxLoop utxoIndex privKeys [] preBalancedTx
+      -- (balancedTx, minUtxos) <- balanceTxLoop utxoIndex privKeys [] preBalancedTx
+      (balancedTx, minUtxos) <- balanceTxLoop utxoIndexS privKeys [] preBalancedTx
+
+      let txOuts1 = txOutputs balancedTx
+          addDatumTxs txos = txos <> Map.elems utxoIndexD
+          addDatums tx = tx {txOutputs = addDatumTxs (txOutputs tx)}
```

Comment:
Wait - why are we adding all the txos as outputs?
utxoIndexD is all the utxos in the input lookups that contain datums. This often contains utxos we're not spending in the transaction.
Also, we usually do want to modify uxto datums?
I'm guessing the logic we'd want here would be any utxos that get _added_ by balancing, must return with the same value and datum - but then in that case, theres no point consuming them.
Could you elaborate on the justification here?
If a script input is consumed and expected to remain the same, that is part of the tx constraints for that tx, as its specific to the project.

Reply:
Okay, I just misunderstood what should be done. I've removed all of the code that splits up utxos by whether they have datums, and only left the (new) code that doesn't add change to UTxOs with datums.